### ISSUE TITLE
add a static keyword to python3

### ISSUE THREAD

**Issue by wolfwood** [InBetweenNames/gentooLTO#439]
[`Fedora Stakeholders Debate Statically Linking Python For Better Performance`](https://www.phoronix.com/scan.php?page=news_item&px=Fedora-Static-Link-Python-F32)

[Fedora Change Proposal](https://fedoraproject.org/wiki/Changes/PythonStaticSpeedup)

basically, python interpreter use a shared lib, libpython3. you can statically link the python interpreter and see speedups on some extremely short lived benchmarks. down side is that it eats more storage space, by a few MB.

static linking seems likely to aid LTO,  so I thought peeps here would be interested.

I'd like to assemble a small set of phoronix python benchmarks to test out this, and our pgo USE flag, ala #431 .

the three line patch [here](https://src.fedoraproject.org/fork/churchyard/rpms/python3/blob/13a0e435b07f8590cb5a911a25de895033772267/f/00666.patch) applies cleanly, but I don't know the best way to conditionally apply this patch. it seems the ebuild currently uses all patches unconditionally.

you can test for static linking by running `ldd /usr/lib/python-exec/python3.6/python3`, but please note that `/usr/bin/python3` is not actually the interpreter, its something that hooks into eselect to dispatch the selected one.

**Comment by jiblime** [comment]
If this does get traction, it would be wise to do some vigorous testing since static libraries and LTO get weird. Even with `-ffat-lto-objects` some packages will crash. There is also the (small) issue of having to rebuild anything statically linked.

**Comment by wolfwood** [comment]
just to be clear, python is already building both a static and a shared
library (though fedora currently patches out the static lib, so they are
also reversing that).

the patch i shared just switched which one the interpreter is linked
against.

**Comment by wolfwood** [comment]
[Quick Python Benchmark of Static linking and PGO](https://openbenchmarking.org/result/1911186-HV-PYTHON18602&obr_shm=y&obr_sgm=y&obr_swl=y&obr_nor=y)

Looks like PGO is a definite win, but static is not an obvious improvement.

I didn't mess with python2 since the static patch didn't apply cleanly, and since EOL is coming, but I guess I should double check that some of the benchmarks aren't using it instead of python3.

**Comment by wolfwood** [comment]
lol, yeah the compile-bench and numpy tests use python2

**Comment by jiblime** [comment]
oof

**Comment by jiblime** [comment]
Interesting update to your [original link](https://fedoraproject.org/wiki/Changes/PythonStaticSpeedup):

> #### Obsoleted by another change
>This change was originally proposed for Fedora 32, but it has been obsoleted by another change proposal: [Build Python with -fno-semantic-interposition for better performance](https://fedoraproject.org/wiki/Changes/PythonNoSemanticInterpositionSpeedup).

>When we build the Python interpreter with the -fno-semantic-interposition compiler/linker flag, we can achieve a performance gain of 5% to 27% depending on the workload. Link time optimizations and profile guided optimizations also have a greater impact when python3 is built this way. 

Looks like @InBetweenNames did a really good job picking out these flags.

There is a lot more information on that page. 

- An explanation of what `-fno-semantic-interposition` can do in tandem with a static binary 

- Benchmarks

- Ways to benchmark Python

- Python sanity testing

**tl;dr** the proposed change has changed from just `static` to `static + -fno-semantic-interposition`. Added bonus is documented proof by a distribution that GentooLTO is [doing Gentoo right](https://www.shlomifish.org/humour/by-others/funroll-loops/Gentoo-is-Rice.html#third).

**Comment by wolfwood** [comment]
ah. [disabling SEMINTERPOS does indeed prove this point](https://openbenchmarking.org/result/1911216-HV-PYTHON92677). LTO + SEMINTERPOS give most of the benefit of static linking.
